### PR TITLE
Youtube related improvements

### PR DESCRIPTION
**Pull request by emutavchi**

### CODE REVIEW

**Comment by eocanha** [comment]
I need at least a detailed explanation of how this new startup buffering mechanism works. From what I could understand, it's a way to bypass the "appsink-new-sample" bus message (which pumps the samples one by one to the main thread), store all the samples retrieved in the appsink thread in an intermediate m_startupSamples Vector field, and process all of them in a row in the main thread when the "need-data" condition arrives. However, it's not clear to me how m_startupDurationMs can become 0 again when a new append operation starts in the future and there are some other loose ends that I still can't understand.

**Comment by emutavchi** [comment]
@eocanha Hi Enrique, the observation here is that during video playback startup YouTube spends a lot of time handling its internal model change (~1-1.5 seconds, sometimes more). So, the idea of "startup" buffering is to continue parsing when UI thread is occupied by JS; and also provide a mechanizm to "flush" parsed samples when some event occures. I.e. with this change we can apply a hack to "flush" active pipelines when some event is occurs, for example when app issues xhr or image request, which YouTube does during the model update.

File: Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp [hunk 1]
```diff
@@ -320,18 +320,29 @@ bool MediaPlayerPrivateGStreamerMSE::doSeek()
...
-        if (setStateResult == GST_STATE_CHANGE_FAILURE) {
-            GST_DEBUG("[Seek] Cannot seek, failed to pause playback pipeline.");
-            webKitMediaSrcSetReadyForSamples(WEBKIT_MEDIA_SRC(m_source.get()), true);
-            m_seeking = false;
-            return false;
+        // Media source may trigger seek completion even when the target time is not yet buffered,
+        // in this case it is better continue the seek and wait for the app to provide media data.
+        m_mseSeekCompleted=true;
+        m_mediaSource->seekToTime(seekTime);
+        if (m_mseSeekCompleted == false)
```
Comment (eocanha) [inline Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp]: Please, use !m_mseSeekCompleted and open braces in this line, both in this block and in the "} else {" one.

File: Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp [hunk 2]
```diff
@@ -256,6 +263,40 @@ void AppendPipeline::dispatchDecryptionKey(GstBuffer* buffer)
 }
 #endif
 
+void AppendPipeline::setStartupBufferingComplete(bool complete)
+{
+    LockHolder locker(m_newSampleLock);
+    m_startupBufferingComplete = complete;
+}
+
+void AppendPipeline::flushStartupSamples()
```
Comment (eocanha) [inline Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp]: please, add ASSERT(isMainThread()) or ASSERT(!isMainThread()) to all the new methods to clarify in which thread they're supposed to be running. Also check the {
  braces
}.

File: Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp [hunk 3]
```diff
@@ -256,6 +263,40 @@ void AppendPipeline::dispatchDecryptionKey(GstBuffer* buffer)
...
+{
+    if (!enableStartupBuffering() || !m_playerPrivate)
+        return;
+
+    Vector< GRefPtr<GstSample> > samples;
+    {
+        LockHolder locker(m_newSampleLock);
```
Comment (eocanha) [inline Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp]: This can be done with locker.unlockEarly() instead of using a code block.

Comment (emutavchi) [reply Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp]: done

File: Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp (hunk 2 above)
Comment (emutavchi) [reply Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp]: done

File: Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp (hunk 1 above)
Comment (emutavchi) [reply Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp]: done

**Pull request merged**